llvm/include/llvm/SandboxIR/Tracker.h

```diff
@@ -0,0 +1,181 @@
+// tracker. IR-change objects are subclasses of `IRChangeBase` and get
+// registered with the `SandboxIRTracker::track()` function. The change objects
+// are saved in the order they are registered with the tracker and are stored in
+// the `SandboxIRTracker::Changes` vector. All of this is done transparently to
+// the user.
+//
+// Reverting changes
+// -----------------
+// Calling `SandboxIRTracker::revert()` will restore the state saved when
+// `SandboxIRTracker::save()` was called. Internally this goes through the
+// change objects in `SandboxIRTracker::Changes` in reverse order, calling their
+// `IRChangeBase::revert()` function one by one.
+//
+// Accepting changes
+// -----------------
+// The user needs to either revert or accept changes before the tracker object
+// is destroyed. This is enforced in the tracker's destructor.
+// This is the job of `SandboxIRTracker::accept()`. Internally this will go
+// through the change objects in `SandboxIRTracker::Changes` in order, calling
+// `IRChangeBase::accept()`.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SANDBOXIR_TRACKER_H
+#define LLVM_SANDBOXIR_TRACKER_H
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Module.h"
+#include "llvm/SandboxIR/Use.h"
+#include "llvm/Support/Debug.h"
+#include <memory>
+#include <regex>
+
+namespace llvm::sandboxir {
+
+class BasicBlock;
+
+/// Each IR change type has an ID.
+enum class TrackID {
+  UseSet,
+};
+
+#ifndef NDEBUG
+static const char *trackIDToStr(TrackID ID) {
+  switch (ID) {
+  case TrackID::UseSet:
+    return "UseSet";
+  }
+  llvm_unreachable("Unimplemented ID");
+}
+#endif // NDEBUG
+
+class SandboxIRTracker;
+
+/// The base class for IR Change classes.
+class IRChangeBase {
+protected:
+#ifndef NDEBUG
+  unsigned Idx = 0;
+#endif
+  const TrackID ID;
+  SandboxIRTracker &Parent;
+
+public:
+  IRChangeBase(TrackID ID, SandboxIRTracker &Parent);
+  TrackID getTrackID() const { return ID; }
+  /// This runs when changes get reverted.
+  virtual void revert() = 0;
+  /// This runs when changes get accepted.
+  virtual void accept() = 0;
+  virtual ~IRChangeBase() = default;
+#ifndef NDEBUG
+  void dumpCommon(raw_ostream &OS) const {
+    OS << Idx << ". " << trackIDToStr(ID);
+  }
+  virtual void dump(raw_ostream &OS) const = 0;
+  LLVM_DUMP_METHOD virtual void dump() const = 0;
+  friend raw_ostream &operator<<(raw_ostream &OS, const IRChangeBase &C) {
+    C.dump(OS);
+    return OS;
+  }
+#endif
+};
+
+/// Tracks the change of the source Value of a sandboxir::Use.
+class UseSet : public IRChangeBase {
+  Use U;
+  Value *OrigV = nullptr;
+
+public:
+  UseSet(const Use &U, SandboxIRTracker &Tracker)
+      : IRChangeBase(TrackID::UseSet, Tracker), U(U), OrigV(U.get()) {}
+  // For isa<> etc.
+  static bool classof(const IRChangeBase *Other) {
```

Comment:
what is this used for? it doesn't really make sense to have both virtual inheritance and `isa<>` on a class

Reply:
Ah I think this is dead code. I think it was meant to be used for fast upcasting, but I think it's not used anywhere. I will remove it. Thanks for noticing.

Reply:
Hmm now that `classof()` is not needed, the only use of the `TrackID` should be for dumping the class name when debugging. This could be done by simply passing the class name to the parent constructor like:
```
  UseSet(const Use &U, Tracker &Tracker) : IRChangeBase("UseSet", &Parent: Tracker), ...
```
But ideally we should only use the name in the debug build. 
What is a good way of #ifdefing it? Should I use two separate constructors? or one constructor with an #ifdefed argument like:
```
 UseSet(const Use &U, Tracker &Tracker) : IRChangeBase(
#ifndef NDEBUG
 "UseSet", 
 #endif // NDEBUG
 &Parent: Tracker), ...
 ```

Reply:
do you expect that this is actually going to get used? can we just drop it?

Reply:
i.e. how often is this actually going to help with debugging an issue? and how often are issues around this going to appear?

Reply:
in general having too many differences between a NDEBUG and non-NDEBUG build makes it more annoying to develop and harder to catch issues that only arise in one configuration. I'd lean toward simplifying when possible and only adding debugging aids when the types of issues it helps debug come up often enough and the debugging aid actually makes it easier to debug

Reply:
When there is a bug in checkpointing it is useful to get a dump and see which changes are in the Changes vector and look if any of them looks suspicious. Without it you will just have to guess.

Reply:
Another option is to keep it in the prod build for now and remove it later once the whole infrastructure is stable.

Reply:
can you manually print the string in the `dump()` override? instead of keeping something as a member variable

Reply:
Yeah, that works too.